pkg/plugins/executable.go

```diff
@@ -0,0 +1,114 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package plugins
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/ghodss/yaml"
+	"sigs.k8s.io/kustomize/pkg/ifc"
+	"sigs.k8s.io/kustomize/pkg/pgmconfig"
+	"sigs.k8s.io/kustomize/pkg/resmap"
+)
+
+type ExecPlugin struct {
+	name string
+	arg  string
+	file string
+	rf   *resmap.Factory
+	ldr  ifc.Loader
+}
+
+func (p *ExecPlugin) Config(
+	ldr ifc.Loader, rf *resmap.Factory, k ifc.Kunstructured) error {
+	dir := filepath.Join(pgmconfig.ConfigRoot(), "plugins")
+	id := k.GetGvk()
+	p.name = filepath.Join(dir, id.Group, id.Version, id.Kind)
+	p.rf = rf
+	p.ldr = ldr
+
+	var err error
+	p.arg, err = k.GetFieldValue("arg")
+	if err != nil && !strings.Contains(err.Error(), "no field named") {
+		return err
+	}
+	p.file, err = k.GetFieldValue("file")
```

Comment:
why are we hard coding these?  the plugin implementations should be reading these, not the plugin manager

Reply:
Since the `stdin` is occupied by resmaps, we can't use `stdin` to configure the executables.

Reply:
An example of a transformer config is given elsewhere here as the file `sed-transformer.yaml` containing

```
apiVersion: someteam.example.com/v1
kind: SedTransformer
metadata:
  name: some-random-name
arg: whatever
file: sed-input.txt
```
We could
* pass this yaml blob to the plugin via stdin, as, say - the first resource before all the resources to be transformed, and add a `---` separator
* pass the file name `sed-transformer.yaml` to the plugin as a command line arg, and assume the plugin will read this arg as a file name, and then be ready to parse the YAML.
* or do what's done here, recognize command line args, and pass them to the plugin as command line args.

The last option is general purpose, since an arg could be a file name, and the file it refers to can be in the kustomization directory, and the plugin exec could go read that file while still accepting resources from stdin.

This option works well out of the box with legacy commands, e.g. sed, helm, whatever.  You don't have to wrap the legacy in some kustomize YAML detangler.